utils/aplay/aplay.c

```diff
@@ -681,20 +796,61 @@ static void *io_worker_routine(struct io_worker *w) {
 		}
 
 		if (!w->alsa_mixer.has_mute_switch && pcm_muted)
-			snd_pcm_format_set_silence(pcm_format, buffer.data, ffb_len_out(&buffer));
+#if ENABLE_RESAMPLER
+			snd_pcm_format_set_silence(resampler_native_format(w->alsa_pcm.format), write_buffer->data, ffb_len_out(write_buffer));
+#else
+			snd_pcm_format_set_silence(w->alsa_pcm.format, write_buffer->data, ffb_len_out(write_buffer));
+#endif
 
-		if (alsa_pcm_write(&w->alsa_pcm, &buffer, !w->ba_pcm.running, verbose) < 0)
+#if ENABLE_APLAY_RESAMPLER
+		if (use_resampler) {
+			if (resampler_process(resampler, &read_buffer, write_buffer) != 0)
+				goto close_alsa;
+		}
+#endif
+
+		if (alsa_pcm_write(&w->alsa_pcm, write_buffer, !w->ba_pcm.running, verbose) < 0)
 			goto close_alsa;
 
+		/* Set poll() timeout such that this thread is always woken before an
+		 * ALSA underrun can occur */
+		timeout = 1000 * w->alsa_pcm.hw_avail / w->alsa_pcm.rate;
+		/* poll() timeouts may be late by up to 2ms depending on the scheduler
+		 * and workload. So we allow for this when setting the timeout value. */
+		timeout -= 2;
```

Comment:
Is it guaranteed that the timeout calculated a line above will be >= 2 (otherwise we will go into negative values)? Also, the `w->alsa_pcm.hw_avail` is updated after the `if (drain) { ... }` block and 0 is returned. In such case the `hw_avail` will not bw updated (jump to `close_alsa` is done only if `alsa_pcm_write` returns < 0).

Reply:
> Is it guaranteed that the timeout ... will be >= 2 ?

No, a fix for that is on its way.

> In such case the hw_avail will not be updated

I will set both the delay and hw_avail values to 0 after a drain, since they do not have any meaningful value when pcm state is SND_PCM_STATE_SETUP. In any case, we only drain when the ba_pcm becomes idle, so I will move the ba_pcm.running check before the timeout setting. New commit hopefully later today after testing.